cl/phase1/network/services/attestation_service.go

```diff
@@ -186,56 +299,77 @@ func (s *attestationService) ProcessMessage(ctx context.Context, subnet *uint64,
-	startSlotAtEpoch := targetEpoch * s.beaconCfg.SlotsPerEpoch
-	if targetBlock := s.forkchoiceStore.Ancestor(root, startSlotAtEpoch); targetBlock != att.AttestantionData().Target().BlockRoot() {
-		return fmt.Errorf("invalid target block. root %v targetEpoch %v attTargetBlockRoot %v targetBlock %v", root.Hex(), targetEpoch, att.AttestantionData().Target().BlockRoot().Hex(), targetBlock.Hex())
-	}
-	// [IGNORE] The current finalized_checkpoint is an ancestor of the block defined by attestation.data.beacon_block_root --
-	// i.e. get_checkpoint_block(store, attestation.data.beacon_block_root, store.finalized_checkpoint.epoch) == store.finalized_checkpoint.root
-	startSlotAtEpoch = s.forkchoiceStore.FinalizedCheckpoint().Epoch() * s.beaconCfg.SlotsPerEpoch
-	if s.forkchoiceStore.Ancestor(root, startSlotAtEpoch) != s.forkchoiceStore.FinalizedCheckpoint().BlockRoot() {
-		return fmt.Errorf("invalid finalized checkpoint %w", ErrIgnore)
-	}
+	// further processing will be done after async signature verification
+	aggregateVerificationData.F = func() {
+		// [IGNORE] The block being voted for (attestation.data.beacon_block_root) has been seen (via both gossip and non-gossip sources)
+		// (a client MAY queue attestations for processing once block is retrieved).
+		if _, ok := s.forkchoiceStore.GetHeader(root); !ok {
+			s.scheduleAttestationForLaterProcessing(att)
+			return
+		}
 
-	err = s.committeeSubscribe.CheckAggregateAttestation(att)
-	if errors.Is(err, aggregation.ErrIsSuperset) {
-		return ErrIgnore
-	}
-	if err != nil {
-		return err
+		// [REJECT] The attestation's target block is an ancestor of the block named in the LMD vote -- i.e.
+		// get_checkpoint_block(store, attestation.data.beacon_block_root, attestation.data.target.epoch) == attestation.data.target.root
+		startSlotAtEpoch := targetEpoch * s.beaconCfg.SlotsPerEpoch
+		if targetBlock := s.forkchoiceStore.Ancestor(root, startSlotAtEpoch); targetBlock != att.Attestation.AttestantionData().Target().BlockRoot() {
+			log.Warn("invalid target block. root ", root.Hex(), " targetEpoch ", targetEpoch, " attTargetBlockRoot ", att.Attestation.AttestantionData().Target().BlockRoot().Hex(), " targetBlock ", targetBlock.Hex())
+			return
+		}
+		// [IGNORE] The current finalized_checkpoint is an ancestor of the block defined by attestation.data.beacon_block_root --
+		// i.e. get_checkpoint_block(store, attestation.data.beacon_block_root, store.finalized_checkpoint.epoch) == store.finalized_checkpoint.root
+		startSlotAtEpoch = s.forkchoiceStore.FinalizedCheckpoint().Epoch() * s.beaconCfg.SlotsPerEpoch
+		if s.forkchoiceStore.Ancestor(root, startSlotAtEpoch) != s.forkchoiceStore.FinalizedCheckpoint().BlockRoot() {
+			log.Warn(fmt.Sprint("invalid finalized checkpoint %w", ErrIgnore))
+			return
+		}
+
+		err = s.committeeSubscribe.CheckAggregateAttestation(att.Attestation)
```

Comment:
you should only batch verify signatures not these ifs

Reply:
Removed ifs from the func() (running after successful verification). But this one remains there because `CheckAggregateAttestation` adds attestation into the pull which we don't want to happen if signature verification fails.